File: examples/sudoku/sudoku.ts
export function solve(boardStr: string): string {
    if (boardStr.length !== 81) {
        return "Invalid board length";
    }

    const board: number[][] = [];
    for (let i = 0; i < 9; i++) {
        const row: number[] = [];
        for (let j = 0; j < 9; j++) {
            const char = boardStr[i * 9 + j];
            row.push(char === '.' ? 0 : parseInt(char, 10));
        }
        board.push(row);
    }

    if (solveSudoku(board)) {
        let result = "";
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                result += board[i][j].toString();
            }
        }
        return result;
    }

    return "Unsolvable";
}

function solveSudoku(board: number[][]): boolean {
    for (let row = 0; row < 9; row++) {
        for (let col = 0; col < 9; col++) {
            if (board[row][col] === 0) {
                for (let num = 1; num <= 9; num++) {
                    if (isValid(board, row, col, num)) {
                        board[row][col] = num;
                        if (solveSudoku(board)) {
                            return true;
                        }
                        board[row][col] = 0;
                    }
                }
                return false;
            }
        }
    }
    return true;
}

function isValid(board: number[][], row: number, col: number, num: number): boolean {
    for (let x = 0; x < 9; x++) {
        if (board[row][x] === num || board[x][col] === num) {
            return false;
        }
    }
    const startRow = Math.floor(row / 3) * 3;
    const startCol = Math.floor(col / 3) * 3;
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            if (board[i + startRow][j + startCol] === num) {
                return false;
            }
        }
    }
    return true;
}

File: examples/python/sudoku/sudoku.py
class WitWorld:
    def solve(self, board_str: str) -> str:
        """Solve a sudoku puzzle given as an 81-char string (. for empty cells)."""
        if len(board_str) != 81:
            return "Invalid board length"

        board: list[list[int]] = []
        for i in range(9):
            row: list[int] = []
            for j in range(9):
                c = board_str[i * 9 + j]
                row.append(0 if c == "." else int(c))
            board.append(row)

        if self._solve(board):
            return "".join(str(board[i][j]) for i in range(9) for j in range(9))
        return "Unsolvable"

    def _solve(self, board: list[list[int]]) -> bool:
        for row in range(9):
            for col in range(9):
                if board[row][col] == 0:
                    for num in range(1, 10):
                        if self._is_valid(board, row, col, num):
                            board[row][col] = num
                            if self._solve(board):
                                return True
                            board[row][col] = 0
                    return False
        return True

    def _is_valid(self, board: list[list[int]], row: int, col: int, num: int) -> bool:
        for x in range(9):
            if board[row][x] == num or board[x][col] == num:
                return False
        sr, sc = (row // 3) * 3, (col // 3) * 3
        for i in range(3):
            for j in range(3):
                if board[sr + i][sc + j] == num:
                    return False
        return True
